libs/ballot-interpreter-nh-next/src/timing_marks.rs

```diff
@@ -0,0 +1,1038 @@
+    pub partial_timing_marks: Partial,
+
+    /// Timing marks inferred from the partial timing marks.
+    pub complete_timing_marks: Complete,
+
+    /// Areas of the ballot card that contain shapes that may be timing marks.
+    pub candidate_timing_marks: Vec<Rect>,
+
+    /// Metadata from the ballot card bottom timing marks.
+    pub metadata: BallotPageMetadata,
+}
+
+impl TimingMarkGrid {
+    pub fn new(
+        geometry: Geometry,
+        orientation: Orientation,
+        border_inset: Inset,
+        scaled_size: Size<u32>,
+        partial_timing_marks: Partial,
+        complete_timing_marks: Complete,
+        candidate_timing_marks: Vec<Rect>,
+        metadata: BallotPageMetadata,
+    ) -> Self {
+        Self {
+            geometry,
+            orientation,
+            border_inset,
+            scaled_size,
+            partial_timing_marks,
+            complete_timing_marks,
+            candidate_timing_marks,
+            metadata,
+        }
+    }
+
+    /// Returns the center of the grid position at the given coordinates. Timing
+    /// marks are at the edges of the grid, and the inside of the grid is where
+    /// the ovals are.
+    ///
+    /// For example, if the grid is 34x51, then:
+    ///
+    ///   - (0, 0) is the top left corner of the grid
+    ///   - (33, 0) is the top right corner of the grid
+    ///   - (0, 50) is the bottom left corner of the grid
+    ///   - (33, 50) is the bottom right corner of the grid
+    ///   - (c, r) where 0 < c < 33 and 0 < r < 50 is the oval at column c and
+    ///     row r
+    pub fn point_for_location(&self, column: u32, row: u32) -> Option<Point<f32>> {
+        if column >= self.geometry.grid_size.width || row >= self.geometry.grid_size.height {
+            return None;
+        }
+
+        let timing_mark_width = self.geometry.timing_mark_size.width.round() as u32;
+        let left = self.complete_timing_marks.left_rects.get(row as usize)?;
+        let right = self.complete_timing_marks.right_rects.get(row as usize)?;
+        let corrected_left = Rect::new(
+            left.right() - timing_mark_width as i32,
+            left.top(),
+            timing_mark_width,
+            left.height(),
+        );
+        let corrected_right =
+            Rect::new(right.left(), right.top(), timing_mark_width, right.height());
+        let horizontal_segment = Segment::new(corrected_left.center(), corrected_right.center());
+        let distance_percentage = column as f32 / (self.geometry.grid_size.width - 1) as f32;
+        let Segment {
+            start: _,
+            end: expected_timing_mark_center,
+        } = horizontal_segment.with_length(horizontal_segment.length() * distance_percentage);
+        Some(expected_timing_mark_center)
+    }
+}
+
+/// Finds the timing marks in the given image and computes the grid of timing
+/// marks, i.e. the locations of all the possible ovals.
+#[time]
+pub fn find_timing_mark_grid(
```

Comment:
Could use some comments explaining what each step in this algorithm does. Specifically I'm not totally clear what "candidate", "partial", "complete", "actual", and "filled" timing marks are and what purposes they serve.

Reply:
Helpful! Thanks